plans/painstall.pp

```diff
@@ -0,0 +1,102 @@
+# Checks for puppet-agent and installs or updates if necessary
+#
+# @param nodes
+#   Target nodeset for puppet-agent
+# @param agent_version
+#   The minimum version of puppet-agent to install or update to
+# @param update
+#   Update existing puppet-agent versions
+
+plan simp_bolt::painstall (
+  TargetSpec $nodes,
+  String     $agent_version = '5.5.14-1',
+  Boolean    $update        = false
+) {
+  # Query nodes for puppet-agent
+  $results = run_task('package::linux', $nodes, name => 'puppet-agent', action => 'status')
+
+  # Generate set of nodes with no puppet-agent
+  $need_agent = $results.filter |$result| { $result[status] == "uninstalled" }
+  $install_subset = $need_agent.map |$result| { $result.target }
+  $ver_inst_results = run_task('facts', $install_subset)
+
+  # Generate set of nodes that require a newer puppet agent
+  $have_agent = $results.filter |$result| { $result[status] == "installed" }
+  $need_update = $have_agent.filter |$result| { versioncmp($result[version], "$agent_version") == -1 }
+  $update_subset = $need_update.map |$result| { $result.target }
+  $ver_upd_results = run_task('facts', $update_subset)
+
+  $r = ["6","7"]
+  $r.each |$r| {
+    # For new installs
+    $rel_installs = $ver_inst_results.filter |$result| { $result['os']['release']['major'] == "${r}" }
+    $install_rel_subset = $rel_installs.map |$result| { $result.target }
+
+    # Check existing repo for adequate version
+    $repo_version = run_task('simp_bolt::payum', $install_rel_subset,
+             version => "${agent_version}.el${r}")
+    # Install agent_version if available from yum
+    $yum_tgt_rel_subset = $repo_version.filter |$result| { versioncmp($result[_output], "${agent_version}.el${r}") == 0 }
+    $install_target_rel_subset = $yum_tgt_rel_subset.map |$result| { $result.target }
+    run_task('package::linux', $install_target_rel_subset,
+             name    => 'puppet-agent',
+             version => "${agent_version}.el${r}",
+             action  => 'install')
+    # Install newer agent_version if available from yum
+    $yum_newer_rel_subset = $repo_version.filter |$result| { versioncmp($result[_output], "$agent_version.el${r}") == 1 }
```

Comment:
I think addition of a max_version parameter would meet this criteria.  We could pass it to the payum script and if the target version is not available, it would return the newest version available to yum that meets or is less than the max version.